File: lib/data/topics.dart
import 'package:flutter/material.dart';

import '../models/topic.dart';

/// The MVP's fixed topic list (PRD §5). Self-selected by the user — no
/// placement test.
const List<Topic> kTopics = [
  Topic(
    id: TopicId.gerundVsInfinitive,
    title: 'Gerund vs. Infinitive',
    description: 'Knowing when it\'s "I enjoy swimming" vs. "I want to swim".',
    // Two forms to choose between for the same idea.
    icon: Icons.compare_arrows,
  ),
  Topic(
    id: TopicId.modalVerbs,
    title: 'Modal Verbs',
    description: 'Can, could, must, should, might, and the nuance between them.',
    // Modals express the speaker's mental stance toward possibility/necessity.
    icon: Icons.psychology,
  ),
  Topic(
    id: TopicId.modalPastForms,
    title: 'Modal Past Forms',
    description: 'Past and reported contexts: will → would, can → could.',
    icon: Icons.history,
  ),
  Topic(
    id: TopicId.tenseSelection,
    title: 'Tense Selection',
    description: 'Choosing the right tense for the situation.',
    icon: Icons.access_time,
  ),
  Topic(
    id: TopicId.articles,
    title: 'Articles',
    description: 'A, an, the — and when to use none at all.',
    icon: Icons.abc,
  ),
];

Topic topicById(TopicId id) => kTopics.firstWhere((t) => t.id == id);

File: proxy/src/topics.ts
/**
 * Mirrors `lib/data/topics.dart`'s `kTopics` exactly (id, title,
 * description). Duplicated here on purpose rather than shared: the worker
 * is a separate deployable with no access to the Dart source, and the
 * client is only ever allowed to send a `topicId` — never a title or
 * description string of its own — so prompt text can't be injected through
 * this field. The worker looks up the real title/description from this
 * fixed table instead of trusting anything the client sends beyond the id.
 */
export interface TopicInfo {
  readonly id: string;
  readonly title: string;
  readonly description: string;
}

export const TOPICS: readonly TopicInfo[] = [
  {
    id: 'gerundVsInfinitive',
    title: 'Gerund vs. Infinitive',
    description: 'Knowing when it\'s "I enjoy swimming" vs. "I want to swim".',
  },
  {
    id: 'modalVerbs',
    title: 'Modal Verbs',
    description: 'Can, could, must, should, might, and the nuance between them.',
  },
  {
    id: 'modalPastForms',
    title: 'Modal Past Forms',
    description: 'Past and reported contexts: will → would, can → could.',
  },
  {
    id: 'tenseSelection',
    title: 'Tense Selection',
    description: 'Choosing the right tense for the situation.',
  },
  {
    id: 'articles',
    title: 'Articles',
    description: 'A, an, the — and when to use none at all.',
  },
];

const BY_ID = new Map(TOPICS.map((t) => [t.id, t]));

export function topicById(id: string): TopicInfo | undefined {
  return BY_ID.get(id);
}

export function isKnownTopicId(id: unknown): id is string {
  return typeof id === 'string' && BY_ID.has(id);
}
